### ISSUE TITLE
Can't serialise Dictionaries

### ISSUE THREAD

**Issue by oxinabox** [tanmaykm/ProtoBuf.jl#69]
Test case:

```
using ProtoBuf
dd = Dict("a"=>3)
iob = PipeBuffer()
writeproto(iob, dd)
```

result:

```
ERROR: UndefRefError: access to undefined reference
 in writeproto(::Base.AbstractIOBuffer{Array{UInt8,1}}, ::Array{String,1}, ::ProtoBuf.ProtoMetaAttribs) at /home/ubuntu/.julia/v0.5/ProtoBuf/src/codec.jl:241
 in writeproto(::Base.AbstractIOBuffer{Array{UInt8,1}}, ::Dict{String,Int64}, ::ProtoBuf.ProtoMeta) at /home/ubuntu/.julia/v0.5/ProtoBuf/src/codec.jl:267
 in writeproto(::Base.AbstractIOBuffer{Array{UInt8,1}}, ::Dict{String,Int64}) at /home/ubuntu/.julia/v0.5/ProtoBuf/src/codec.jl:263
 in eval(::Module, ::Any) at ./boot.jl:234
 in macro expansion at ./REPL.jl:92 [inlined]
 in (::Base.REPL.##1#2{Base.REPL.REPLBackend})() at ./event.jl:46
```

Because:

```
julia> dd.keys
16-element Array{String,1}:
 #undef
 #undef
 #undef
 #undef
 #undef
 #undef
 #undef
 #undef
 #undef
 #undef
 #undef
 #undef
 #undef
    "a"
 #undef
 #undef
```

So the undefined elements are not being handled -- any attempt to read them will result in  a `UndefRefError`

**Comment by tanmaykm** [comment]
Yes, as you point out, arrays in the Julia `Dict` data structure can contain undefined entries. They can not be serialized are they are with protobuf.

An alternative could be to extract valid keys and values and serialize. E.g.:

````
julia> type SerDict
           k::Array{String}
           v::Array{Int}
       end
julia> sd = SerDict(collect(keys(d)), collect(values(d)));
julia> iob = PipeBuffer();
julia> writeproto(iob, sd)
0x0000000000000005
````

**Comment by tanmaykm** [comment]
The alternative I mentioned is similar to how protobuf `map` types are serialized now.
I'm also working on supporting `Dict` with `map`, and hope to have it out soon.

**Comment by TotalVerb** [comment]
I am not familiar with ProtoBuf, but is there a technological reason that `undef` cannot be serialized? The definedness of an element can be determined with `isdefined(A, i)`, and deserialization can be done by allocating a vector and only assigning the defined items.

**Comment by tanmaykm** [comment]
@TotalVerb Yes, `#undef`s can be determined.
But the protobuf spec does not have a way to send only specific (non contiguous) elements of an array.

**Comment by TotalVerb** [comment]
How does protobuf handle pointer types, like strings? Does it not have a null pointer value? (that's what `#undef` is anyway).

**Comment by TotalVerb** [comment]
If there is no null pointer value, it should be possible to define a nullable container type for all non-`isbits` values.

**Comment by tanmaykm** [comment]
Types supported by protobuf are specified here: https://developers.google.com/protocol-buffers/docs/proto

**Comment by TotalVerb** [comment]
Seems like this should be possible:

```
message StringRef {
    optional StringObject _ = 1
}

message StringObject {
    ...
}
```

Any type that is not `isbits` can have the `Ref` type generated. This does introduce some overhead.

**Comment by tanmaykm** [comment]
ProtoBuf `map` types are now mapped to `Dict`.

Fixed via #70
